src/Aspire.Hosting.Azure.KeyVault/AzureKeyVaultResourceExtensions.cs

```diff
@@ -32,23 +34,42 @@ public static IResourceBuilder<AzureKeyVaultResource> AddAzureKeyVault(this IDis
-    /// <param name="configureResource">Callback to configure the underlying <see cref="global::Azure.Provisioning.KeyVaults.KeyVault"/> resource.</param>
+    /// <param name="configureResource">Callback to configure the underlying <see cref="global::Azure.Provisioning.KeyVault.KeyVaultService"/> resource.</param>
     /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
     [Experimental("AZPROVISION001", UrlFormat = "https://aka.ms/dotnet/aspire/diagnostics#{0}")]
-    public static IResourceBuilder<AzureKeyVaultResource> AddAzureKeyVault(this IDistributedApplicationBuilder builder, string name, Action<IResourceBuilder<AzureKeyVaultResource>, ResourceModuleConstruct, KeyVault>? configureResource)
+    public static IResourceBuilder<AzureKeyVaultResource> AddAzureKeyVault(this IDistributedApplicationBuilder builder, string name, Action<IResourceBuilder<AzureKeyVaultResource>, ResourceModuleConstruct, KeyVaultService>? configureResource)
     {
         builder.AddAzureProvisioning();
 
         var configureConstruct = (ResourceModuleConstruct construct) =>
         {
-            var keyVault = construct.AddKeyVault(name: construct.Resource.Name);
-            keyVault.AddOutput("vaultUri", x => x.Properties.VaultUri);
+            var keyVault = new KeyVaultService(construct.Resource.Name)
+            {
+                Properties = new KeyVaultProperties()
+            };
+            construct.Add(keyVault);
 
-            keyVault.Properties.Tags["aspire-resource-name"] = construct.Resource.Name;
+            construct.Add(new BicepOutput("vaultUri", typeof(string))
+            {
+                Value =
+                    new MemberExpression(
+                        new MemberExpression(
+                            new IdentifierExpression(keyVault.ResourceName),
+                            "properties"),
+                        "vaultUri")
+            });
 
-            var keyVaultAdministratorRoleAssignment = keyVault.AssignRole(RoleDefinition.KeyVaultAdministrator);
-            keyVaultAdministratorRoleAssignment.AssignProperty(x => x.PrincipalId, construct.PrincipalIdParameter);
-            keyVaultAdministratorRoleAssignment.AssignProperty(x => x.PrincipalType, construct.PrincipalTypeParameter);
+            keyVault.Tags["aspire-resource-name"] = construct.Resource.Name;
+
+            var role = KeyVaultBuiltInRole.KeyVaultAdministrator;
+            var roleAssignment = new RoleAssignment($"{KeyVaultBuiltInRole.GetBuiltInRoleName(role)}_{keyVault.ResourceName}")
+            {
+                Scope = new IdentifierExpression(keyVault.ResourceName),
+                PrincipalType = construct.PrincipalTypeParameter,
+                RoleDefinitionId = BicepFunction.GetSubscriptionResourceId("Microsoft.Authorization/roleDefinitions", role.ToString()),
+                PrincipalId = construct.PrincipalIdParameter
+            };
+            construct.Add(roleAssignment);
```

Comment:
Is this really the new API? These are fine low level building blocks bug oh man...

Reply:
There is an `AssignRole` API, but it assumes a user assigned identity.

https://github.com/Azure/azure-sdk-for-net/blob/ade2cee32561d57a3049d584b17ab749a0bb64c7/sdk/provisioning/Azure.Provisioning.KeyVault/src/Generated/KeyVaultService.cs#L157-L171

```C#
    /// <summary>
    /// Assign a role to a user-assigned identity that grants access to this
    /// KeyVaultService.
    /// </summary>
    /// <param name="role">The role to grant.</param>
    /// <param name="identity">The <see cref="UserAssignedIdentity"/>.</param>
    /// <returns>The <see cref="RoleAssignment"/>.</returns>
    public RoleAssignment AssignRole(KeyVaultBuiltInRole role, UserAssignedIdentity identity) =>
        new($"{identity.ResourceName}_{KeyVaultBuiltInRole.GetBuiltInRoleName(role)}_{ResourceName}")
        {
            Scope = new IdentifierExpression(ResourceName),
            PrincipalType = RoleManagementPrincipalType.ServicePrincipal,
            RoleDefinitionId = BicepFunction.GetSubscriptionResourceId("Microsoft.Authorization/roleDefinitions", role.ToString()),
            PrincipalId = identity.PrincipalId
        };
```

Here we take in a `PrincipleType` as a parameter, so this higher level API doesn't seem to meet our needs.

Reply:
Yes - this is a known point in time thing.  We're starting with the lowest level layer to make sure you're able to do anything.  We want to add more conveniences that provide reasonable starting points over time kind of like https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/provisioning/Azure.Provisioning.Storage/src/StorageResources.cs#L35

Reply:
While not solving the larger problem, I added an `AssignRole(role, principalType, principalId)` overload in https://github.com/Azure/azure-sdk-for-net/pull/45911

Reply:
Thanks @tg-msft !